### PR TITLE
fix: Incorrect lazy CSV `select(len())` for compressed files

### PR DESCRIPTION
**Pull request by nameexhaustion**
Fixes https://github.com/pola-rs/polars/issues/18057
Fixes https://github.com/pola-rs/polars/issues/18070

### CODE REVIEW

File: crates/polars-io/src/utils/other.rs [hunk 1]
```diff
@@ -47,9 +47,9 @@ pub fn get_reader_bytes<'a, R: Read + MmapBytesReader + ?Sized>(
 ///
 /// # Safety
 /// The `out` vec outlives `bytes` (declare `out` first).
-pub unsafe fn maybe_decompress_bytes<'a>(
+pub fn maybe_decompress_bytes<'a, 'b: 'a>(
```
Comment (nameexhaustion) [inline crates/polars-io/src/utils/other.rs:50]: learned new things about lifetimes recently 😁

File: crates/polars-io/src/csv/read/parser.rs [hunk 2]
```diff
@@ -41,13 +41,25 @@ pub fn count_rows(
...
+    let mmap = unsafe { memmap::Mmap::map(&file).unwrap() };
+    let owned = &mut vec![];
+    let mut reader_bytes = maybe_decompress_bytes(mmap.as_ref(), owned)?;
+
+    for _ in 0..reader_bytes.len() {
+        if reader_bytes[0] != eol_char {
+            break;
+        }
+
+        reader_bytes = &reader_bytes[1..];
```
Comment (nameexhaustion) [inline crates/polars-io/src/csv/read/parser.rs:54]: skip empty lines at the start

File: crates/polars-io/src/utils/other.rs (hunk 1 above)
Comment (ritchie46) [reply crates/polars-io/src/utils/other.rs:47]: Then we can remove the `# Safety` comment as well. :+1:

Comment (nameexhaustion) [reply crates/polars-io/src/utils/other.rs:47]: Ah, I removed - it doesn't show unless you go into "Files changed"

**Review by ritchie46 — APPROVED** [review]

File: crates/polars-io/src/utils/other.rs (hunk 1 above)
Comment (TraoreB1-j) [reply crates/polars-io/src/utils/other.rs]: Format 
Page

**Pull request merged**